Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,142 @@
+import enum
+
+from types import TracebackType
+from typing import final, Any, Dict, Optional, Type
+
+from . import events
+from . import tasks
+
+
+__all__ = (
+    "Timeout",
+    "timeout",
+    "timeout_at",
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    CANCELLING = "cancelling"
+    CANCELLED = "cancelled"
+    EXITED = "exited"
+
+
+@final
+class Timeout:
+
+    def __init__(self, deadline: Optional[float]) -> None:
+        self._state = _State.CREATED
+
+        self._timeout_handler: Optional[events.TimerHandle] = None
+        self._task: Optional[tasks.Task[Any]] = None
+        self._deadline = deadline
+
+    @property
+    def deadline(self) -> Optional[float]:
+        return self._deadline
+
+    def reschedule(self, deadline: Optional[float]) -> None:
+        assert self._state != _State.CREATED
+        if self._state != _State.ENTERED:
+            raise RuntimeError(
+                f"Cannot change state of {self._state} CancelScope",
+            )
+
+        self._deadline = deadline
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+
+        if deadline is None:
+            self._timeout_handler = None
+        else:
+            loop = events.get_running_loop()
+            self._timeout_handler = loop.call_at(
+                deadline,
+                self._on_timeout,
+            )
+
+    def expired(self) -> bool:
```

Comment:
Method, like `task.cancelled()`.
`.expired` is better name if the timeout context

Reply:
+1